### PR TITLE
marksman: init at 2022-12-28

### LINKED ISSUE #184873: Packaging Request: marksman
**Project description**
Markdown LSP server providing completion, cross-references, diagnostics, and more. 

**Metadata**

* homepage URL: https://github.com/artempyanykh/marksman
* source URL: https://github.com/artempyanykh/marksman/commit/f9bc6139566418130f5f269b75f359cefc9a492d
* license: mit
* platforms: linux, darwin

### PR DESCRIPTION
**Pull request by stasjok**
###### Description of changes

Language server for Markdown: https://github.com/artempyanykh/marksman.

Closes #184873.

I plan to use it myself and thought that I can try to add it to nixpkgs. Never used dotnet, but based on my basic testing, the build is successful and the app is working (tested on linux).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/marksman/default.nix
Comment (TLATER) [inline pkgs/development/tools/marksman/default.nix:7]: Nit: If possible, it would really be better to use the new [recursive attributes feature of `mkDerivation`](https://nixos.org/manual/nixpkgs/stable/#idm140737319947216), but I don't know if `buildDotnetModule` supports that yet.

File: pkgs/development/tools/marksman/default.nix
Comment (TLATER) [inline pkgs/development/tools/marksman/default.nix:42]: Are you planning to contribute this package without a maintainer? Or are you simply unsure as to whether it's ok to add yourself to the contributor list?

File: pkgs/development/tools/marksman/default.nix
Comment (TLATER) [inline pkgs/development/tools/marksman/default.nix:25]: Since this is a relatively complex update process, it would be really nice if there were an [update script](https://nixos.org/manual/nixpkgs/stable/#var-passthru-updateScript) for this one, especially if you don't intend to maintain it yourself.

I've never used dotnet either, but I imagine it wouldn't be too hard to do the nuget fetching in a script?

File: pkgs/development/tools/marksman/default.nix
Comment (TLATER) [inline pkgs/development/tools/marksman/default.nix:41]: MIT requires:

> The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

Currently this isn't copied into the package:

```console
tlater ~/result $ ls
bin  lib
tlater ~/result $ fd LICENSE .
```

Arguably a problem with their build scripts, I suppose, but let's copy it in, given their license asks for it? Something like

```nix
postInstall = ''
  mkdir -p $out/share/doc
  install -m 0644 LICENSE $out/share/doc/
'';
```

This isn't always upheld in nixpkgs, but it's a tiny amount of effort, so...

**Review by TLATER — APPROVED** [review]
Looks good to me and works fine too, at a quick use. I left some comments, but it looks mostly fine to me when compared to existing nixpkgs packages :)

File: pkgs/development/tools/marksman/default.nix
Comment (stasjok) [reply pkgs/development/tools/marksman/default.nix:9]: > but I don't know if `buildDotnetModule` supports that yet.

I don't think it does.

**Comment by stasjok** [comment]
I've added updateScript and a very simple test.

**Comment by TLATER** [comment]
Final note, the commits need to be squashed, I'll add it to the reviewed-prs post once that's done.

**Comment by stasjok** [comment]
> Final note, the commits need to be squashed, I'll add it to the reviewed-prs post once that's done.

Done.

**Comment by TLATER** [comment]
Result of `nixpkgs-review pr 193682` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>marksman</li>
  </ul>
</details>

**Review by TLATER — APPROVED** [review]

File: pkgs/development/tools/marksman/default.nix
Comment (SuperSandro2000) [reply pkgs/development/tools/marksman/default.nix:9]: Why would we want to use that here? Just to substitute version in the fetcher? The fetcher is FOD anyway, so it would only work when taking extra care to overwrite the fetcher which is not documented anywhere. More the opposite is that src is always overwritten in all examples. Also people tend to use e.g. pname all over the place which would catastrophically fail with finalAttrs.

File: pkgs/development/tools/marksman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/marksman/default.nix:11]:
```suggestion
  version = "unstable-2022-10-30";
```

File: pkgs/development/tools/marksman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/marksman/default.nix:20]:
```suggestion
  patches = [ ./0001-revert-embed-git-commit-into-the-version.patch ];
```

File: pkgs/development/tools/marksman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/marksman/default.nix:35]: This will conflict with any other program that puts a file there and create needless conflicts.

```suggestion
```

File: pkgs/development/tools/marksman/update.sh
Comment (SuperSandro2000) [inline pkgs/development/tools/marksman/update.sh:9]:
```suggestion
old_version=$(nix-instantiate --eval -A marksman.version | jq -r)
```

File: pkgs/development/tools/marksman/update.sh
Comment (SuperSandro2000) [inline pkgs/development/tools/marksman/update.sh:6]:
```suggestion
```
We are not setting this in many update scripts so I am a bit reluctant to do it here.

File: pkgs/development/tools/marksman/update.sh
Comment (SuperSandro2000) [inline pkgs/development/tools/marksman/update.sh:8]: This should optionally take a GITHUB_TOKEN

```suggestion
version=$(curl -s ${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""} https://api.github.com/repos/artempyanykh/marksman/releases/latest | jq -r .tag_name)
```

File: pkgs/development/tools/marksman/default.nix
Comment (SuperSandro2000) [reply pkgs/development/tools/marksman/default.nix:56]: meta.maintainers must be set otherwise we cannot accept the PR.

File: pkgs/development/tools/marksman/default.nix
Comment (SuperSandro2000) [reply pkgs/development/tools/marksman/default.nix:55]: > This isn't always upheld in nixpkgs

Yes, because it creates conflicts between packages when installed globally.

File: pkgs/development/tools/marksman/default.nix
Comment (TLATER) [reply pkgs/development/tools/marksman/default.nix:35]: How about

```
postInstall = ''
    install -m 644 -Dt $out/share/doc/marksman/ LICENSE
'';
```

instead then? At least on my system a few packages seem to already be placing license files along with documentation there, and this matches what other distros do.

Comment (stasjok) [reply pkgs/development/tools/marksman/default.nix:35]: Yes, if we want to put a LICENSE, it'll have to be `$out/share/doc/marksman`. I didn't notice that it's in common dir, my bad.

File: pkgs/development/tools/marksman/default.nix
Comment (stasjok) [reply pkgs/development/tools/marksman/default.nix:11]: I didn't manage to build it with this version.

1. Version starting with a string is not allowed in dotnet.
1. Looks like dotnet is using `Version` or `version` env to lookup a version, or a tag from project file. But in a build phase nix overrides even that:
https://github.com/NixOS/nixpkgs/blob/b457130e8a21608675ddf12c7d85227b22a27112/pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-build-hook.sh#L23-L25

And to be honest I don't like it much. I know about the rule:

> If a package is not a release but a commit from a repository, then the version attribute must be the date of that (fetched) commit. The date must be in `unstable-YYYY-MM-DD` format.

But it's not a random commit, it's a tagged release on github which happens to be a date. But yes, it looks like there is no traditional versioning for this project yet.

Any suggestions?

**Comment by stasjok** [comment]
I believe I addressed most of the conversations.

File: pkgs/development/tools/marksman/default.nix
Comment (TLATER) [reply pkgs/development/tools/marksman/default.nix:11]: `dotnetBuildFlags` seem to come after that version flag, does dotnet allow overriding arguments so you can hard-code your own version flag to override the one the build phase sets?

File: pkgs/development/tools/marksman/update.sh
Comment (stasjok) [reply pkgs/development/tools/marksman/update.sh]: Just a note. I noticed this `${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""}` in some places in nixpkgs, but I think it's just wrong. This substitution expands to one word: `-u ":$GITHUB_TOKEN"`. Correct substitution `${GITHUB_TOKEN:+-u ":$GITHUB_TOKEN"}` expands to two words: `-u` and `:$GITHUB_TOKEN`.

You can test it:

```bash
stas@server2:~$ a="aa bb"
stas@server2:~$ set -x
stas@server2:~$ : ${a:+"-u \":$a\""}
+ : '-u ":aa bb"'
stas@server2:~$ : ${a:+-u ":$a"}
+ : -u ':aa bb'
```

In essence with `${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""}` curl will use username ` "` (space and quote) and password `$GITHUB_TOKEN"` (with quote at the end). Token will be invalid.

File: pkgs/development/tools/marksman/default.nix
Comment (stasjok) [reply pkgs/development/tools/marksman/default.nix:11]: I already tried that without luck. I tested it again and my problem was with configure phase. With `dotnetFlags` it works. But it somewhat complicates things, especially with update script.

Comment (stasjok) [reply pkgs/development/tools/marksman/default.nix]: OK. Even if I don't like it, since no one is answering, I've added `unstable-` prefix. I hope it can be merged now.

File: pkgs/development/tools/marksman/update.sh
Comment (SuperSandro2000) [reply pkgs/development/tools/marksman/update.sh]: Can you fix this up, please?

``GITHUB_API_TOKEN`` is just another alias for ``GITHUB_TOKEN`` and probably the 5th I have in my configuration.

File: maintainers/maintainer-list.nix
Comment (SuperSandro2000) [inline maintainers/maintainer-list.nix:12984]: Please add github and githubId.

File: pkgs/development/tools/marksman/default.nix
Comment (SuperSandro2000) [inline pkgs/development/tools/marksman/default.nix:12]:
```suggestion
  version = "2022-10-30";
```
there was a confusion here. I thought this was an unstable tag.

File: pkgs/development/tools/marksman/update.sh
Comment (SuperSandro2000) [reply pkgs/development/tools/marksman/update.sh]: I just did a double check and it seems to expand correct. I think we are missing an outer quote, so it is resplit.

**Comment by michaeladler** [comment]
Nice work (once upton a time I tried to package as well but got too frustrated with the whole dotnet packaging experience)! I've cherry-picked your commit and the LSP works fine in neovim.

File: pkgs/development/tools/marksman/update.sh
Comment (SuperSandro2000) [reply pkgs/development/tools/marksman/update.sh]: > You still think that `${GITHUB_TOKEN:+"-u \":$GITHUB_TOKEN\""}` is correct?

No, but it works probably due to resplicing.

Comment (SuperSandro2000) [reply pkgs/development/tools/marksman/update.sh]: #204219

**Comment by edrex** [comment]
There's a new release https://github.com/artempyanykh/marksman/releases/tag/2022-11-25
Better to bump before merge?

**Comment by edrex** [comment]
New major feature release: https://github.com/artempyanykh/marksman/releases/tag/2022-12-19

**Review by avdv — APPROVED** [review]

**Comment by avdv** [comment]
This seems good to go. Anything left to do here?

**Comment by stasjok** [comment]
> New major feature release: https://github.com/artempyanykh/marksman/releases/tag/2022-12-19

I just pushed the update.

> This seems good to go. Anything left to do here?

No. It's ready.

**Comment by avdv** [comment]
@stasjok thanks for the update! :+1: 

Could you take another look @SuperSandro2000 ? Thanks!

**Comment by stasjok** [comment]
1. Updated to **2022-12-23**.
2. Dropped a patch that prevented running git during build process. Thanks to artempyanykh/marksman#125.

Now the version string reported by `marksman --version` is `1.0.0-2022-12-23` (upstream binaries have version like `1.0.0-662682b`). `1.0.0` is still hard-coded here: https://github.com/artempyanykh/marksman/blob/662682b9d5fc163ed4be3ce0dba54febd6258bf3/Marksman/Marksman.fsproj#L7

@SuperSandro2000, what do you think if we change a version of the package to `1.0.0-<git tag>` too? I don't think that `1.0.0` has any meaning upstream, it's like a placeholder for version (at least for now).

**Review by pacien — APPROVED** [review]
Imported and tested on 22.11.
It is automatically detected and used by the Helix editor.
Everything seems to be working fine.

**Comment by SuperSandro2000** [comment]
If a software tags releases we use those tags as the version number and strip prefixes like v, r or release.

**Pull request merged**
